cms/djangoapps/modulestore_migrator/tasks.py

```diff
@@ -0,0 +1,923 @@
+    # We use this mapping to ensure that we don't create duplicate
+    # PublishableEntities during the migration process for a given LearningPackage.
+    existing_source_to_target_keys = {
+        block.source.key: block.target for block in ModulestoreBlockMigration.objects.filter(
+        overall_migration__target=migration.target.id)
+    }
+
+    migration_context = MigrationContext(
+        existing_source_to_target_keys=existing_source_to_target_keys,
+        target_package_id=target_package_pk,
+        target_library_key=target_library.library_key,
+        source_context_key=source_root_usage_key.course_key,
+        composition_level=CompositionLevel(composition_level),
+        replace_existing=replace_existing,
+        preserve_url_slugs=preserve_url_slugs,
+        created_by=status.user_id,
+        created_at=datetime.now(timezone.utc),
+    )
+
+    with authoring_api.bulk_draft_changes_for(migration.target.id) as change_log:
+        root_migrated_node = _migrate_node(
+            context=migration_context,
+            content_by_filename=content_by_filename,
+            source_node=root_node,
+        )
+    change_log.save()
+    migration.change_log = change_log
+    status.increment_completed_steps()
+
+    status.set_state(MigrationStep.UNSTAGING.value)
+    staged_content.delete()
+    status.increment_completed_steps()
+
+    _create_migration_artifacts_incrementally(
+        root_migrated_node=root_migrated_node,
+        source=source,
+        migration=migration,
+        status=status,
+    )
+
+    block_migrations = ModulestoreBlockMigration.objects.filter(overall_migration=migration)
+    status.increment_completed_steps()
+
+    status.set_state(MigrationStep.FORWARDING.value)
+    if forward_source_to_target:
+        block_sources_to_block_migrations = {
+            block_migration.source: block_migration for block_migration in block_migrations
+        }
+        for block_source, block_migration in block_sources_to_block_migrations.items():
+            block_source.forwarded = block_migration
+            block_source.save()
+        # ModulestoreBlockSource.objects.bulk_update(block_sources_to_block_migrations.keys(), ["forwarded"])
+        source.forwarded = migration
+        source.save()
+    status.increment_completed_steps()
+
+    status.set_state(MigrationStep.POPULATING_COLLECTION.value)
+    if target_collection:
+        block_target_pks: list[int] = list(
+            ModulestoreBlockMigration.objects.filter(
+                overall_migration=migration
+            ).values_list('target_id', flat=True)
+        )
+        if block_target_pks:
+            log.info("Started adding entities to a collection")
+            # @@TODO: Investigate an issue with huge
+            # time consumption during this step
+            # For existing and deleted blocks this step takes
+            # up to 110 seconds
+            # For new blocks this step takes up to 3 seconds
```

Comment:
Can you write a GitHub issue for this and link it here?

Reply:
The TODO is removed - can't reproduce the issue.